### PR TITLE
feat: change the logic to handle dynamic batching

### PR DESCRIPTION
**Pull request by JoanFM**
This PR fixes the behavior of dynamic batching.

- It makes sure that no batch larger than preferred batch size gets into the Executor.
- It allows the requests to be returned to the client as they are processed without waiting for other requests to be ready, this should improve the average latency of the system.

### CODE REVIEW

File: jina/serve/runtimes/worker/batch_queue.py [hunk 1]
```diff
@@ -97,73 +99,143 @@ async def push(self, request: DataRequest) -> Task:
...
+        return queue
 
     async def _await_then_flush(self) -> None:
         """Process all requests in the queue once flush_trigger event is set."""
-        await self._flush_trigger.wait()
 
-        # writes to shared data between tasks need to be mutually exclusive
-        async with self._data_lock:
-            try:
+        def _get_buckets_completed_request_indexes(non_assigned_docs,
```
Comment (JohannesMessner) [inline jina/serve/runtimes/worker/batch_queue.py:118]: It is not super clear to me what this method does. what does one "bucket" represent?

File: jina/serve/runtimes/worker/batch_queue.py [hunk 2]
```diff
@@ -82,7 +84,7 @@ async def _sleep_then_set(self, event: Event):
         await asyncio.sleep(self._timeout / 1000)
         event.set()
 
-    async def push(self, request: DataRequest) -> Task:
+    async def push(self, request: DataRequest) -> asyncio.Queue:
         """Append request to the queue. Once the request has been processed, the returned task will complete.
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply jina/serve/runtimes/worker/batch_queue.py]: Done

File: jina/serve/runtimes/worker/batch_queue.py [hunk 3]
```diff
@@ -97,73 +99,143 @@ async def push(self, request: DataRequest) -> Task:
                 self._flush_task = asyncio.create_task(self._await_then_flush())
             if not self._timer_task:
                 self._start_timer()
-
             self._big_doc.extend(docs)
-            self._requests.append(request)
+            nex_req_idx = len(self._requests)
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply jina/serve/runtimes/worker/batch_queue.py]: Done

File: jina/serve/runtimes/worker/batch_queue.py [hunk 4]
```diff
@@ -97,73 +102,173 @@ async def push(self, request: DataRequest) -> Task:
...
-        await self._flush_trigger.wait()
 
+        def _get_docs_groups_completed_request_indexes(non_assigned_docs,
+                                                       non_assigned_docs_reqs_idx,
+                                                       sum_from_previous_mini_batch_in_first_req_idx):
+            """
+            This method groups all the `non_assigned_docs` into groups of docs according to the `req_idx` they belong to.
+            They are only distributed when we are sure that the request is full.
+
+            The method returns the ac
```
Comment (JohannesMessner) [inline jina/serve/runtimes/worker/batch_queue.py:128]: ac?

**Review by JohannesMessner — DISMISSED** [review]
just one tiny comment

**Comment by JohannesMessner** [comment]
btw i think it is fair to call this a feature or at least a performance improvement, the old implementation was correct (we never made any guarantees that the actual batch size would not be bigger than the preferred one), just not optimized for this scenario.

**Pull request merged**